csi/moac/pool_operator.ts

```diff
@@ -0,0 +1,476 @@
+      }
+      disks = disks.slice(0).sort();
+      //if (typeof disks !== 'string') {
+      this.spec = { node, disks };
+    }
+    this.status = {
+      state: poolStateFromString(cr.status?.state),
+      reason: cr.status?.reason,
+      disks: cr.status?.disks,
+      capacity: cr.status?.capacity,
+      used: cr.status?.used,
+    };
+  }
+
+  // Extract name of the pool from the resource metadata.
+  getName(): string {
+    if (this.metadata.name === undefined) {
+      throw Error("Resource object does not have a name")
+    } else {
+      return this.metadata.name;
+    }
+  }
+}
+
+// Pool operator tries to bring the real state of storage pools on mayastor
+// nodes in sync with mayastorpool custom resources in k8s.
+export class PoolOperator {
+  namespace: string;
+  watcher: CustomResourceCache<PoolResource>; // k8s resource watcher for pools
+  registry: any; // registry containing info about mayastor nodes
+  eventStream: any; // A stream of node and pool events.
+  workq: any; // for serializing pool operations
+
+  // Create pool operator.
+  //
+  // @param namespace     Namespace the operator should operate on.
+  // @param kubeConfig    KubeConfig.
+  // @param registry      Registry with node objects.
+  // @param [idleTimeout] Timeout for restarting watcher connection when idle.
+  constructor (
+    namespace: string,
+    kubeConfig: KubeConfig,
+    registry: any,
+    idleTimeout: number | undefined,
+  ) {
+    this.namespace = namespace;
+    this.registry = registry; // registry containing info about mayastor nodes
+    this.eventStream = null; // A stream of node and pool events.
+    this.workq = new Workq(); // for serializing pool operations
+    this.watcher = new CustomResourceCache(
+      this.namespace,
+      RESOURCE_NAME,
+      kubeConfig,
+      PoolResource,
+      { idleTimeout }
+    );
+  }
+
+  // Create pool CRD if it doesn't exist.
+  //
+  // @param kubeConfig  KubeConfig.
+  async init (kubeConfig: KubeConfig) {
+    log.info('Initializing pool operator');
+    let k8sExtApi = kubeConfig.makeApiClient(ApiextensionsV1Api);
+    try {
+      await k8sExtApi.createCustomResourceDefinition(crdPool);
+      log.info(`Created CRD ${RESOURCE_NAME}`);
+    } catch (err) {
+      // API returns a 409 Conflict if CRD already exists.
+      if (err.statusCode !== 409) throw err;
+    }
+  }
+
+  // Start pool operator's watcher loop.
+  //
+  // NOTE: Not getting the start sequence right can have catastrophic
+  // consequence leading to unintended pool destruction and data loss
+  // (i.e. when node info is available before the pool CRD is).
+  //
+  // The right order of steps is:
+  //   1. Get pool resources
+  //   2. Get info about pools on storage nodes
+  async start () {
+    var self = this;
+
+    // get pool k8s resources for initial synchronization and install
+    // event handlers to follow changes to them.
+    self._bindWatcher(self.watcher);
+    await self.watcher.start();
```

Comment:
yes. this is the right way to do it (not the other way). watcher starts to emit events on behalf of start so it must be bound before we start it otherwise we could miss events.